core/utils/wasm-builder/src/wasm_project.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{fs, path::{Path, PathBuf}, borrow::ToOwned, process::Command};
+
+use toml::value::Table;
+
+/// The name for the wasm binary.
+const WASM_BINARY: &str = "wasm_binary";
+
+/// Creates the WASM project, compiles the WASM binary and compacts the WASM binary.
+///
+/// # Returns
+/// The path to the compact WASM binary.
+pub fn create_and_compile(cargo_manifest: &Path) -> PathBuf {
+	let project = create_project(cargo_manifest);
+
+	build_project(&project);
+	compact_wasm_file(&project);
+
+	project.join(format!("{}.compact.wasm", WASM_BINARY))
+}
+
+/// Find the `Cargo.lock` relative to the `OUT_DIR`.
+fn find_cargo_lock() -> PathBuf {
+	let mut out_dir = build_helper::out_dir();
+
+	while out_dir.pop() {
+		if out_dir.join("Cargo.lock").exists() {
+			return out_dir.join("Cargo.lock")
+		}
+	}
+
+	panic!("Did not found `Cargo.lock` for out dir: {}", build_helper::out_dir().display())
+}
+
+/// Extract the crate name from the given `Cargo.toml`.
+fn get_crate_name(cargo_manifest: &Path) -> String {
```

Comment:
If this is called on the current crate, you can use the `CARGO_PKG_NAME` environment variable.

Reply:
Good idea, will need to try that.

Reply:
Okay, does not work.